### ISSUE TITLE
vscode-server broken

### ISSUE THREAD

**Issue by steffen86** [microsoft/vscode#147695]
Issue Type: <b>Bug</b>

Since I updated to the latest version of Visual Studio Code (1.66.2), the vscode-server used by the remote ssh extension is not working anymore. VSCode connects over ssh to the server and then downloads the vscode-server but when it's started, it failes with node: not found error.

I logged into the server via ssh and tried to start ~/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/node manually with the same error. I looked around and found an other version of node in a subfolder with a different hash. This can be started thus I assume that something is wrong with the node file.

Here is the output from Visual Studio Code:

```
[15:33:46.704] Log Level: 2
[15:33:46.706] remote-ssh@0.78.0
[15:33:46.707] win32 x64
[15:33:46.769] SSH Resolver called for "ssh-remote+rpi4.fritz.box", attempt 1
[15:33:46.770] "remote.SSH.useLocalServer": false
[15:33:46.771] "remote.SSH.showLoginTerminal": false
[15:33:46.771] "remote.SSH.remotePlatform": {"rpi4.fritz.box":"linux"}
[15:33:46.772] "remote.SSH.path": undefined
[15:33:46.772] "remote.SSH.configFile": undefined
[15:33:46.772] "remote.SSH.useFlock": true
[15:33:46.772] "remote.SSH.lockfilesInTmp": false
[15:33:46.773] "remote.SSH.localServerDownload": auto
[15:33:46.773] "remote.SSH.remoteServerListenOnSocket": false
[15:33:46.773] "remote.SSH.showLoginTerminal": false
[15:33:46.773] "remote.SSH.defaultExtensions": []
[15:33:46.774] "remote.SSH.loglevel": 2
[15:33:46.774] "remote.SSH.enableDynamicForwarding": true
[15:33:46.774] "remote.SSH.enableRemoteCommand": false
[15:33:46.775] "remote.SSH.serverPickPortsFromRange": {}
[15:33:46.775] "remote.SSH.serverInstallPath": {}
[15:33:46.922] SSH Resolver called for host: rpi4.fritz.box
[15:33:46.922] Setting up SSH remote "rpi4.fritz.box"
[15:33:46.960] Using commit id "dfd34e8260c270da74b5c2d86d61aee4b6d56977" and quality "stable" for server
[15:33:46.967] Install and start server if needed
[15:33:46.981] Checking ssh with "ssh -V"
[15:33:47.404] > OpenSSH_for_Wind
[15:33:47.406] > ows_8.1p1, LibreSSL 3.0.2

[15:33:47.430] Running script with connection command: ssh -T -D 65272 "rpi4.fritz.box" bash
[15:33:47.436] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[15:33:48.334] > ]0;C:\WINDOWS\System32\cmd.exe
[15:33:48.335] Got some output, clearing connection timeout
[15:33:48.696] > xxx@rpi4.fritz.box's password: 
[15:33:48.697] Showing password prompt
[15:33:55.510] Got password response
[15:33:55.511] "install" wrote data to terminal: "**********"
[15:33:55.512] > xxx@rpi4.fritz.box's password:
> 
> 
[15:33:55.512] Showing password prompt
[15:34:15.492] "install" terminal command done
[15:34:15.493] Install terminal quit with output: xxx@rpi4.fritz.box's password:
[15:34:15.493] Received install output: xxx@rpi4.fritz.box's password:
[15:34:15.494] Failed to parse remote port from server output
[15:34:15.495] Resolver error: Error: 
	at Function.Create (c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:586041)
	at Object.t.handleInstallOutput (c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:584693)
	at Object.t.tryInstall (c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:681924)
	at processTicksAndRejections (node:internal/process/task_queues:96:5)
	at async c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:644564
	at async Object.t.withShowDetailsEvent (c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:647909)
	at async Object.t.resolve (c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:645643)
	at async c:\Users\z0045srs\.vscode\extensions\ms-vscode-remote.remote-ssh-0.78.0\out\extension.js:1:722950
[15:34:15.513] ------




[15:34:21.500] Password dialog canceled
[15:34:21.501] "install" terminal command canceled
[15:34:21.502] > 
[15:34:21.503] > 3856f23430c6: running
[15:34:21.504] > Acquiring lock on /home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/vscode-remote-lock.xxx.dfd34e8260c270da74b5c2d86d61aee4b6d56977
[15:34:21.505] > Installing to /home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977...
> 3856f23430c6%%1%%
[15:34:21.506] > Downloading with wget
[15:34:21.507] > Download complete
[15:34:21.508] > 3856f23430c6%%2%%
> tar --version:
> tar (GNU tar) 1.34
> Copyright © 2021 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> 
> Geschrieben von John Gilmore und Jay Fenlason.
[15:34:21.509] > Checking /home/xxx/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.log and /home/xxx/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.pid for a running server
[15:34:21.510] > Starting server with command... /home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/bin/code-server --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all   &> "/home/stef
> fen/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.log" < /dev/null
> printenv:
[15:34:21.511] >     SHELL=/bin/bash
>     PWD=/home/xxx
>     LOGNAME=xxx
>     XDG_SESSION_TYPE=tty
>     MOTD_SHOWN=pam
>     HOME=/home/xxx
>     LANG=de_DE.UTF-8
>     VSCODE_AGENT_FOLDER=/home/xxx/.vscode-server
>     SSH_CONNECTION=192.168.178.39 65276 192.168.178.107 22
>     XDG_SESSION_CLASS=user
>     USER=xxx
>     SHLVL=0
>     XDG_SESSION_ID=1553
>     XDG_RUNTIME_DIR=/run/user/1000
>     SSH_CLIENT=192.168.178.39 65276 22
>     PATH=/usr/local/bin:/usr/bin:/bin:/usr/games
>     OLDPWD=/home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977
>     _=/usr/bin/printenv
> rm: das Entfernen von '/home/xxx/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.token' ist nicht möglich: Datei oder Verzeichnis nicht gefunden
> Spawned remote server: 20128
[15:34:21.514] > Waiting for server log...
> Waiting for server log...
[15:34:21.515] > Waiting for server log...
[15:34:21.517] > Waiting for server log...

[15:34:21.725] > Waiting for server log...
[15:34:21.726] >  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
> 
[15:34:21.727] > Server did not start successfully. Full server log at /home/xxx/.vscode-server/.dfd34e8260c270da74b5c2d86d61aee4b6d56977.log >>>
> /home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/bin/code-server: 12: /home/xxx/.vscode-server/bin/dfd34e8260c270da74b5c2d86d61aee4b6d56977/node: not found
> <<< End of server log
[15:34:21.729] > 
> 3856f23430c6: start
> exitCode==32==
> listeningOn====
> osReleaseId==raspbian==
[15:34:21.730] > 
> arch==aarch64==
> tmpDir==/run/user/1000==
> platform==linux==
> unpackResult==success==
> didLocalDownload==0==
> downloadTime==7220==
> installTime==2494==
> extInstallTime====
> serverStartTime==8088==
> 3856f23430c6: end
```

VS Code version: Code 1.66.2 (dfd34e8260c270da74b5c2d86d61aee4b6d56977, 2022-04-11T07:46:01.075Z)
OS version: Windows_NT x64 10.0.19042
Restricted Mode: No
Connection to 'SSH' could not be established 

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz (12 x 2592)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.74GB (4.71GB free)|
|Process Argv|--crash-reporter-id cead2e0a-fc09-4ed7-95dd-835c62fc6bcd|
|Screen Reader|no|
|VM|0%|

Connection to 'SSH' could not be established 
</details>Extensions: none<details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30443607
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
pythonvsnew555:30457759
vscscmwlcmt:30465135

```

</details>

**Comment by stmudie** [comment]
I'm having the same issue. And when I downgraded to 1.65.2 the problem persists even though I was successfully running remote containers up until the upgrade to 1.66.2

**Comment by stmudie** [comment]
I got this to work again by deleting all my containers and, importantly, the `vscode` docker volume.

**Comment by steffen86** [comment]
My setup is different. I am running a raspberry pi to which I connect from a Windows 10 PC. 

I did some more Tests, downloaded older Versions of VSCode which then installs other Versions of the vscode-server. Even the older Versions do not work anymore.

My guess is, that it has to do with the architecture, the remote is running. It seems as if the vscode-server does not fit. So the question: how does VSCode decide which vscode-server to install and can I influence this?

**Comment by stmudie** [comment]
Ah yes I see that you are using remote ssh and I am using remote containers. Interesting that we saw the same error though. My setup is Windows 10 to Linux VM on a remote host, docker via ssh. The remote container extension stores the vscode-server in a docker volume (named vscode) on the remote host, which is then mounted into the remote development container.

The vscode-server version is linked directly to the build you are using of code locally. Hence as you saw multiple versions are being installed. Can you delete all the vscode-servers that have been installed on your RPi, and force it to copy them again?

**Comment by steffen86** [comment]
I had some time to dig deeper...

When first installing the vscode-server on the remote host, the server is downloaded. There is some "magic" to decide what version to download, especially which architecture to download.

I figured out, that it is downloading server-linux-arm64/stable but on my raspberry pi I would need server-linux-armhf/stable. I manually downloaded the correct version and replaced the wrong one. --> It works again.

I also have an explaination why it downloaded the right version earlier and now downloads the "wrong" one.

Recently i enabled an Raspberry Pi option in /boot/config.txt that is called arm_64bit. For my limited understanding this enables 64bit features of the kernel while the system software can remain 32bit. I had to enable it for docker which then is able to run container based on a 64bit system.

So i assume that the "magic" deciding on the server architecture to download looks at the kernel and not at the system and its libraries to distinguish between 32 and 64 bit.

Can I force vscode to download the armhf version instead of the arm64 one?
